src/lib/datepicker/datepicker-input.ts

```diff
@@ -73,20 +99,52 @@ export class MdDatepickerInput<D> implements AfterContentInit, ControlValueAcces
   }
 
   /** The minimum valid date. */
-  @Input() min: D;
+  @Input()
+  get min(): D { return this._min; }
+  set min(value: D) {
+    this._min = value;
+    this._validatorOnChange();
+  }
+  private _min: D;
 
   /** The maximum valid date. */
-  @Input() max: D;
+  @Input()
+  get max(): D { return this._max; }
+  set max(value: D) {
+    this._max = value;
+    this._validatorOnChange();
+  }
+  private _max: D;
 
   /** Emits when the value changes (either due to user input or programmatic change). */
   _valueChange = new EventEmitter<D>();
 
-  _onChange = (value: any) => {};
-
   _onTouched = () => {};
 
+  private _cvaOnChange = (value: any) => {};
+
+  private _validatorOnChange = () => {};
+
   private _datepickerSubscription: Subscription;
 
+  /** The form control validator for this input. */
+  private _validator: ValidatorFn = Validators.compose([
+    (control: AbstractControl): ValidationErrors | null => {
+      return (!this.min || !control.value ||
+          this._dateAdapter.compareDate(this.min, control.value) < 0) ?
+          null : {'mdDatepickerMin': {'min': this.min, 'actual': control.value}};
+    },
+    (control: AbstractControl): ValidationErrors | null => {
+      return (!this.max || !control.value ||
+          this._dateAdapter.compareDate(this.max, control.value) > 0) ?
+          null : {'mdDatepickerMax': {'max': this.max, 'actual': control.value}};
+    },
+    (control: AbstractControl): ValidationErrors | null => {
+      return !this._dateFilter || !control.value || this._dateFilter(control.value) ?
+          null : {'mdDatepickerFilter': true};
+    },
```

Comment:
Have you thought about breaking these validator functions out into their own methods or getters? This is a bit hard to scan.